### DISCUSSION TITLE
Minimum Length Scale Filter Usage Example

### DISCUSSION THREAD

**Discussion by Andeloth** [NanoComp/meep#2115] [Q&A] [answered]
In the recent [meep paper](https://doi.org/10.1364/OE.442074) in Figure 3. it says different minimum length scale filters were used to optimize the following devices.

![image](https://user-images.githubusercontent.com/37753156/175382314-1273b5df-e1a3-44f0-b579-dc5aee8fed16.png)

Where in the repo is the code that was used to produce this result? Alternatively, if the data isn't in the repo, then what filter function was used to enforce the minimum length constraint?

For some context, I'm using the conic filter function to enforce an aspect ratio constraint, aka sidewalls should have a height : width ratio of 3:1, which corresponds to a radius parameter of **0.33...** Does the conic filter also include minimum length filtering, or does another function fulfill that role? I'm asking because the `get_conic_radius_from_eta_e` function seems to indicate to me that the minimum length and the conic filter radius are connected. Or perhaps I am misunderstanding the purpose of the conic filter and the conic filter supposed to be the minimum length filter and etch aspect ratio constraints should be implemented differently?

Thanks!

**Comment by Andeloth** [comment, marked as answer]
I figured it out, after a few hours of searching. Conic filters can enforce minimum length scale constraints, but not aspect ratio constraints. Those I'll just have to implement as a parameter check before optimizations.

**Reply by smartalecH** [reply]
You might be interested in our [other paper](https://doi.org/10.1364/OE.431188) that describes these constraints in detail (also cited in the original meep adjoint paper).

**Reply by Andeloth** [reply]
@smartalecH from what I understand from the paper you linked, the conic/cylindrical/gaussian filters together with the projections are only for enclosed area filtering, and not minimum linewidth and line spacing filtering. Do you have any tutorials or examples that show how to use the indicator and constraint functions in filter.py for linewidth and line spacing filtering?

**Reply by smartalecH** [reply]
> the conic/cylindrical/gaussian filters together with the projections are only for enclosed area filtering, and not minimum linewidth and line spacing filtering.

You have it backward. The filtering (combined with the constraint functions) is precisely for enforcing minimum linewidth and line spacing (i.e. lengthscales)

**Reply by Andeloth** [reply]
I see, then what purpose do the indicator/constraint functions serve? I haven't seen them used in any of the adjoint tutorials, only the conic filter and the projections.

And how would you do enclosed area filtering?

**Reply by Andeloth** [reply]
I guess a more precise question would be, what is the difference between using

1. filters + projections (as shown in adjoint tutorials)
2. filters + constraints (+ projections)

**Reply by smartalecH** [reply]
> I haven't seen them used in any of the adjoint tutorials

We haven't posted any tutorials that properly implement lengthscale constraints, unfortunately.

> I guess a more precise question would be, what is the difference between using...

Density-TO methods *always* use filtering + projection. If you want to enforce a lengthscale, then you *also* need to implement optimization constraints (which in our case, work in tandem with the filtering and projection operations).

> And how would you do enclosed area filtering?

I think you mean to ask, "how do you implement minimum enclosed area constraints?" See #2024.

I would read through that paper a few times just to capture all the naunces. There are a few references we cite, [like this one](https://doi.org/10.1007/s00158-013-0978-6), which give a good overview of the different TO methodologies.

**Reply by Andeloth** [reply]
I see, thanks for the detailed response Alec :D !
